[Smart Accounts Kit ] add delegations buyer guide for x402 (#2912)
MetaMask/metamask-docs: smart-accounts-kit/guides/x402/buyer/advanced-permissions.md

```ts
import { base as chain } from 'viem/chains'

const facilitators = accepted.extra.facilitators
if (!facilitators || facilitators.length === 0) {
  console.error('No facilitators found in PAYMENT-REQUIRED')
  // Handle error
}

const currentTime = Math.floor(Date.now() / 1000)
const expiry = currentTime + 3600

const grantedPermissions = await walletClient.requestExecutionPermissions([
  {
    chainId: chain.id,
    expiry,
    to: sessionAccount.address,
    permission: {
      type: 'erc20-token-allowance',
      data: {
        tokenAddress: accepted.asset,
        // Fixed allowance for this resource.
        allowanceAmount: BigInt(accepted.amount),
        justification: 'Permission to pay for a specific x402-protected API resource',
      },
      isAdjustmentAllowed: false,
    },
    rules: [
      {
        type: 'redeemers',
        data: {
          addresses: facilitators!,
        },
      },
    ],
  },
])
```

### 5. Create a redelegation

The granted advanced permission is delegated to the session account. To let facilitator addresses

> redeem this permission context for x402 settlement, create an open redelegation from the session account.

Reviewer: ```suggestion
redeem this permission context for x402 settlement, create an <GlossaryTerm term="Open redelegation">open redelegation</GlossaryTerm> from the session account.
```

---

> description: Access an x402-protected API data using Advanced Permissions with a fixed allowance.

Reviewer 2: ```suggestion
description: Pay for an x402-protected API data using Advanced Permissions with a fixed allowance.
```

sidebar_label: Advanced Permissions
keywords:
  [x402, ERC-7715, ERC-7710, advanced permissions, redeemers, allowance, session account, buyer]
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import GlossaryTerm from '@theme/GlossaryTerm';

# Pay for an x402 API with Advanced Permissions

In this guide, you request <GlossaryTerm term="Advanced Permissions" /> with a fixed ERC-20 allowance
to pay for a specific x402-protected resource.

## Prerequisites

- [Install and set up the Smart Accounts Kit.](../../../get-started/install.md)

## Steps

### 1. Set up a Wallet Client

Set up a Wallet Client using Viem's [`createWalletClient`](https://viem.sh/docs/clients/wallet) function. This client will

> help you interact with MetaMask.

Reviewer 2: ```suggestion
Set up a Wallet Client using Viem's [`createWalletClient`](https://viem.sh/docs/clients/wallet) function.
Use this client to interact with MetaMask.
```

Then, extend the Wallet Client functionality using `erc7715ProviderActions`.

> These actions enable you to request <GlossaryTerm term="Advanced Permissions" /> from the user.

Reviewer 2: ```suggestion
Extend the Wallet Client with `erc7715ProviderActions` to enable
<GlossaryTerm term="Advanced Permissions" /> requests.
```

```typescript
import { createWalletClient, custom } from 'viem'
import { erc7715ProviderActions } from '@metamask/smart-accounts-kit/actions'

const walletClient = createWalletClient({
  transport: custom(window.ethereum),
}).extend(erc7715ProviderActions())
```

### 2. Set up a session account

> Set up a session account. The requested permissions are granted to the session account, which is responsible for making x402 API calls.

Reviewer 2: What do we mean by session account. May need to clarify.

> The session account can be either a smart account or an EOA. This examples uses EOA as a session account.

Reviewer 2: ```suggestion
The session account can be either a smart account or an Externally owned account (EOA). This examples uses EOA as a session account.
```

Reviewer: Alternatively, use the GlossaryTerm for EOA.
